third_party/move/move-vm/types/src/module_id_interner.rs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::interner::BTreeInterner;
+use move_core_types::language_storage::ModuleId;
+use parking_lot::RwLock;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+pub struct InternedModuleId(pub usize);
+
+pub struct InternedModuleIdPool {
+    inner: RwLock<BTreeInterner<ModuleId>>,
```

Comment:
I think we should fold synchronisation into interner itself? Rwlock does not scale too well for many threads when contended, so we might want to do dashmap / sharding or some tricks to avoid locks for hot ids.